jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,313 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Levenberg-Marquardt algorithm in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+
+from dataclasses import dataclass
+
+from typing_extensions import Literal
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import linear_solve
+from jaxopt._src.linear_solve import solve_cg
+from jaxopt._src.linear_solve import solve_bicgstab
+from jaxopt._src.linear_solve import solve_gmres
+from jaxopt._src.linear_solve import solve_cholesky
+from jaxopt._src.tree_util import tree_l2_norm, tree_inf_norm, tree_sub, tree_add, tree_mul
+from jax.experimental import host_callback as hcb
+
+
+class LevenbergMarquardtState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  damping_coeff: float
+  nu: float
+  residual: Any
+  loss: Any
+  delta: Any
+  error: float
+  dparams_error: float
+  grad_error: float
+  gradient: Any
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class LevenbergMarquardt(base.IterativeSolver):
+  """Gauss-Newton nonlinear least-squares solver.
+
+  Given the residual function ``f(x): R^m -> R^n``, where ``f(x) =
+  residual_fun(x, *args, **kwargs)``, ``LevenbergMarquardt`` finds a local
+  minimum of the cost function ``argmin_x 0.5 * sum(f(x) ** 2)``.
+
+  Attributes:
+    residual_fun: a smooth function of the form
+      ``residual_fun(x, *args, **kwargs)``.
+    maxiter: maximum number of iterations.
+    tol: tolerance.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+  """
+  residual_fun: Callable
+
+  maxiter: int = 30
+  tol: float = 0.0
+
+  damping_parameter: float = 1e-6
+  xtol: float = 1e-8
+  gtol: float = 1e-8
```

Comment:
I am not sure if that would be possible, if we would like to keep the algorithm as is. The algorithm for this implementation suggests stopping either if gradients are small or the delta_params is small. If we have only one tol, we wouldn't be able to distinguish between the two, right? Does that address your question?

The drawback of the current usage is that tol needs to be zero in order to preserve the stopping criteria of the algorithm, at least this was the only way i could think of that doesn't require changing the `_cond_fun` for the `state.error` in the `base.py`.

Reply:
So far, all solvers had a single stopping criterion, typically the l2 norm of  optimality_fun (the gradient in your case). Since optimality_fun is also used for implicit_diff, I think this makes sense and would be my preference for simplicity and consistency.

If you really want to keep , we would indeed need a way to override `_cond_fun` in base.py, as I'm not a fan of using `tol = 0`.

Reply:
I chose to do something in between: I override the `_cond_fun` now but give the user the freedom to choose the `stop_criteria`. WDYT?

Reply:
This works for me. In the future, we may want to provide a public API to do that. This implementation will be part of JAXopt so it's fine to override a private method but for external implementations, it's not ok.